File: 400/467 - Unique Substrings in Wraparound String.c
#include <string.h>

int findSubstringInWraproundString(char* p) {
    int count[26] = {0}, ret = 0, prev_len, i, c;
    
    if (! (p && *p))
        return 0;
    
    c = *p++ - 'a';
    count[c] = 1;
    prev_len = 1;
    
    for (; *p; ++p) {
        if ((c + 1) % 26 == *p - 'a')
            ++prev_len;
        else
            prev_len = 1;

        if (prev_len > count[*p - 'a'])
            count[*p - 'a'] = prev_len;
        
        c = *p - 'a';
    }
    
    for (i = 0; i < 26; ++i)
        ret += count[i];
    
    return ret;
}

File: csharp/467.unique-substrings-in-wraparound-string.cs
/*
 * @lc app=leetcode id=467 lang=csharp
 *
 * [467] Unique Substrings in Wraparound String
 */

using System.Collections.Generic;

public class Solution {
    public int FindSubstringInWraproundString(string p) {
        int len = 0, ret = 0, n = p.Length;
        int[] max = new int[26];
        for (int i = 0; i < n; ++i) {
            int curr = (int) p[i], prev = i > 0 ? (int) p[i - 1] : 0;
            if (i > 0 && curr == prev + 1 || curr == prev - 25) {
                ++len;
            } else {
                len = 1;
            }
            max[curr - 'a'] = Math.Max(len, max[curr - 'a']);
        }
        foreach (int num in max) {
            ret += num;
        }
        return ret;
    }
}

